pkg/middleware/rbac_test.go

```diff
@@ -0,0 +1,554 @@
+	"github.com/labstack/echo/v4"
+	echo_middleware "github.com/labstack/echo/v4/middleware"
+	"github.com/redhatinsights/platform-go-middlewares/identity"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func mockXRhUserIdentity(t *testing.T, org_id string, accNumber string) string {
+	var (
+		err       error
+		xrhid     identity.XRHID
+		jsonBytes []byte
+	)
+	xrhid.Identity.OrgID = org_id
+	xrhid.Identity.AccountNumber = accNumber
+	xrhid.Identity.Internal.OrgID = org_id
+
+	jsonBytes, err = json.Marshal(xrhid)
+	require.NoError(t, err)
+
+	return b64.StdEncoding.EncodeToString([]byte(jsonBytes))
+}
+
+func hasToCallMock(resource string, verb client.RbacVerb, rbacAllowed bool, rbacError error) bool {
+	if resource == "" && verb == "" && rbacAllowed == false && rbacError == nil {
+		return false
+	}
+	return true
+}
+
+func handleItWorked(c echo.Context) error {
+	switch c.Request().Method {
+	case http.MethodGet:
+		return c.JSON(http.StatusOK, "It worked")
+	default:
+		return c.JSON(http.StatusOK, nil)
+	}
+}
+
+func rbacServe(t *testing.T, req *http.Request, resource string, verb client.RbacVerb, skipper echo_middleware.Skipper, rbacAllowed bool, rbacError error, generateIdentity bool) *httptest.ResponseRecorder {
+	var (
+		xrhid string
+		rw    *httptest.ResponseRecorder
+	)
+
+	require.NotNil(t, req)
+
+	if generateIdentity {
+		xrhid = mockXRhUserIdentity(t, "12345", "12345")
+		require.NotEqual(t, "", xrhid)
+	}
+
+	mockRbacClient := mocks_client.NewRbac(t)
+	require.NotNil(t, mockRbacClient)
+	if hasToCallMock(resource, verb, rbacAllowed, rbacError) {
+		mockRbacClient.On("Allowed", xrhid, resource, verb).Return(rbacAllowed, rbacError)
+	}
+
+	e := echo.New()
+	require.NotNil(t, e)
+
+	// Add Rbac middleware
+	e.Use(
+		NewRbac(Rbac{
+			BaseUrl:        config.Get().Clients.RbacBaseUrl,
+			Skipper:        skipper,
+			PermissionsMap: ServicePermissions,
+			Client:         mockRbacClient,
+		}),
+	)
+
+	// Add a handler to avoid 404
+	e.Add(req.Method, req.URL.Path, handleItWorked)
+
+	rw = httptest.NewRecorder()
+	require.NotNil(t, rw)
+
+	if generateIdentity {
+		req.Header.Set(xrhidHeader, xrhid)
+	}
+
+	e.ServeHTTP(rw, req)
+	mockRbacClient.AssertExpectations(t)
+
+	return rw
+}
+
+func skipperTrue(c echo.Context) bool {
+	return true
+}
+
+func skipperFalse(c echo.Context) bool {
+	return false
+}
+
+func TestNewRbacPanics(t *testing.T) {
+	var pm *PermissionsMap = ServicePermissions
+	var client = mocks_client.NewRbac(t)
+	var skipper echo_middleware.Skipper = skipperTrue
+	require.Panics(t, func() {
+		NewRbac(Rbac{
+			BaseUrl:        "http://localhost:8800/api/rbac/v1",
+			Skipper:        skipper,
+			PermissionsMap: nil,
+			Client:         client,
+		})
+	})
+	require.Panics(t, func() {
+		NewRbac(Rbac{
+			BaseUrl:        "http://localhost:8800/api/rbac/v1",
+			Skipper:        skipper,
+			PermissionsMap: pm,
+			Client:         nil,
+		})
+	})
+}
```

Comment:
Could we either add a startup check or a unit test that checks that all registered routes have something defined in ServicePermissions ?  

This way a dev won't ever forget to add a new route to ServicePermissions.

Reply:
I like the idea, and I tried that, unfortunately I didn't find a way (I don't know if it exists) to do the check into the reverse direction; happy to know about it, as that would provide a powerful check detecting that situations in a very early stage (before push the commits) that is very interesting.